libraries/devsec_ssh.rb

```diff
@@ -0,0 +1,220 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module DevSec
+  # Class methods of this class can be called in order to
+  # return different cryptographic configuration data. The calls
+  # can be done based on the CRYPTO constant below. There is autodetection
+  # of required ssh version. If ssh package can not be found, preconfigured OS
+  # matrix is used.
+  #
+  # Syntax:
+  #
+  #   get_[client|server]_[kexs|macs|ciphers] (node-object, enable-weak-ciphers)
+  #
+  # Example:
+  #
+  #   DevSec::Ssh.get_server_macs(node, false)
+  #   => 'hmac-ripemd160 hmac-sha1' # for openssh server version 5.3
+  #
+  class Ssh # rubocop:disable Metrics/ClassLength
+    # Fallback ssh version for autodetection
+    FALLBACK_SSH_VERSION ||= 5.9
+    # Support types of ssh
+    SSH_TYPES ||= [:client, :server].freeze
+    # Crypto configuration for different ssh parameters
+    CRYPTO ||= {
+      kexs: {
+        5.3 => [],
+        5.9 => %w(diffie-hellman-group-exchange-sha256),
+        6.6 => %w(curve25519-sha256@libssh.org diffie-hellman-group-exchange-sha256),
+        :weak => %w(diffie-hellman-group14-sha1 diffie-hellman-group-exchange-sha1 diffie-hellman-group1-sha1)
+      },
+      macs: {
+        5.3 => %w(hmac-ripemd160 hmac-sha1),
+        5.9 => %w(hmac-sha2-512 hmac-sha2-256 hmac-ripemd160),
+        6.6 => %w(hmac-sha2-512-etm@openssh.com hmac-sha2-256-etm@openssh.com
+                  hmac-ripemd160-etm@openssh.com umac-128-etm@openssh.com
+                  hmac-sha2-512 hmac-sha2-256 hmac-ripemd160),
+        :weak => %w(hmac-sha1)
+      },
+      ciphers: {
+        5.3 => %w(aes256-ctr aes192-ctr aes128-ctr),
+        6.6 => %w(chacha20-poly1305@openssh.com aes256-gcm@openssh.com aes128-gcm@openssh.com
+                  aes256-ctr aes192-ctr aes128-ctr),
+        :weak => %w(aes256-cbc aes192-cbc aes128-cbc)
+      }
+    }.freeze
+    # Privilege separation values
+    PRIVILEGE_SEPARATION ||= {
+      5.3 => 'yes',
+      5.9 => 'sandbox'
+    }.freeze
+
+    class << self
+      def get_server_privilege_separarion(node)
+        Chef::Log.debug('Called get_server_privilege_separarion')
+
+        found_ssh_version = find_ssh_version(get_ssh_server_version(node), PRIVILEGE_SEPARATION.keys)
+
+        ret = PRIVILEGE_SEPARATION[found_ssh_version]
+
+        Chef::Log.debug("Using configuration for ssh version #{found_ssh_version}, value: #{ret}")
+
+        ret
+      end
+
+      private
+
+      def method_missing(method, node, enable_weak = false)
+        super unless respond_to_missing?(method)
```

Comment:
Do you mean something like this?

```ruby
def get_data(type, args)
end

def get_client_macs(args)
  get_data('client_macs', args)
end

def get_server_macs(args)
  get_data('server_macs', args)
end
```

My basic idea was to avoid the same code 5 times...